src/Interpreters/OptimizeShardingKeyRewriteInVisitor.cpp

```diff
@@ -45,7 +52,12 @@ bool shardContains(
     if (sharding_column_value.isNull())
         return false;
 
-    Field sharding_value = executeFunctionOnField(sharding_column_value, sharding_column_name, expr, sharding_key_column_name);
+    Field sharding_value = executeFunctionOnField(sharding_column_value, sharding_column_name, sharding_expr, sharding_key_column_name);
+    /// The value from IN can be non-numeric,
+    /// but in this case it should be convertible to numeric type, let's try.
+    if (!isInt64FieldType(sharding_value.getType()))
+        sharding_value = convertFieldToType(sharding_value, DataTypeUInt64());
```

Comment:
1. We can call it unconditionally.
2. Check if it's not convertible (in this case NULL field is returned).

Reply:
>We can call it unconditionally.

Ok

>Check if it's not convertible (in this case NULL field is returned).

convertFieldToType will throw if the conversion is not possible (and there are some tests for this cases)